src/mono/mono/utils/mono-threads-posix.c

```diff
@@ -282,6 +282,15 @@ mono_native_thread_set_name (MonoNativeThreadId tid, const char *name)
 	if (tid != mono_native_thread_id_get ())
 		return;
 
+#if defined(__linux__)
+	/*
+	 * Ignore requests to set the main thread name because
+	 * it causes the value returned by Process.ProcessName to change.
+	 */
+	if (mono_native_thread_os_id_get () == (guint64)getpid ())
+		return;
```

Comment:
This is under `#ifdef __MACH__`.
And not aware of the thread we're trying to set the name for (`MonoNativeThreadId tid` argument).